### PR TITLE
Replace the underlying representation of SingleCliffordGate by CliffordTableau

### PR DESCRIPTION
**Pull request by ybc1991**
The current implementation of SingleQubitCliffordGate is based on `rotation_map` and `inverse_rotation_map`. It is concise for SingleQubitCliffordGate but hard to extend into any number of qubits clifford gates. For generalization, this PR replaces them by the `CliffordTableau`.  All functionalities and interfaces of SingleCliffordGate should be the same after the replacement. [WIP for #3639].

I didn't deprecate PauliTransform (#4088) in this PR for minimize the PR's responsibility. But after this PR, it will be easy to do it.

### CODE REVIEW

File: cirq-core/cirq/ops/clifford_gate.py [hunk 1]
```diff
@@ -43,6 +43,36 @@ def _to_pauli_transform(matrix: np.ndarray) -> Optional[PauliTransform]:
     return None
 
 
+def _to_cliiford_tableau(
```
Comment (balopat) [inline cirq-core/cirq/ops/clifford_gate.py:46]:
```suggestion
def _to_clifford_tableau(
```

File: cirq-core/cirq/ops/clifford_gate.py [hunk 2]
```diff
@@ -68,11 +98,20 @@ class SingleQubitCliffordGate(gate_features.SingleQubitGate):
...
-        self._rotation_map = _rotation_map
-        self._inverse_map = _inverse_map
+        self._clifford_tableau = _clifford_tableau
+
+    @property
+    def clifford_tableau(self):
+        return self._clifford_tableau
+
+    @staticmethod
+    def from_cliiford_tableau(table: qis.CliffordTableau) -> 'SingleQubitCliffordGate':
```
Comment (balopat) [inline cirq-core/cirq/ops/clifford_gate.py:110]:
```suggestion
    def from_clifford_tableau(table: qis.CliffordTableau) -> 'SingleQubitCliffordGate':
```

Comment (balopat) [inline cirq-core/cirq/ops/clifford_gate.py:110]:
```suggestion
    def from_cliiford_tableau(tableau: qis.CliffordTableau) -> 'SingleQubitCliffordGate':
```

File: cirq-core/cirq/ops/clifford_gate.py [hunk 3]
```diff
@@ -237,10 +277,24 @@ def from_unitary(u: np.ndarray) -> Optional['SingleQubitCliffordGate']:
...
 
     def transform(self, pauli: Pauli) -> PauliTransform:
-        return self._rotation_map[pauli]
+        x_to = self._clifford_tableau.destabilizers()[0]
+        z_to = self._clifford_tableau.stabilizers()[0]
+        if pauli == pauli_gates.X:
+            to = x_to
+        elif pauli == pauli_gates.Z:
+            to = z_to
+        else:
```
Comment (balopat) [inline cirq-core/cirq/ops/clifford_gate.py:291]: where is the fourth branch? Can't it be identity?

File: cirq-core/cirq/ops/clifford_gate.py [hunk 4]
```diff
@@ -237,10 +277,24 @@ def from_unitary(u: np.ndarray) -> Optional['SingleQubitCliffordGate']:
...
+            to = x_to * z_to  # Y = iXZ
+            to.coefficient *= 1j
+        single_pauli_gates = (pauli_gates.X, pauli_gates.Y, pauli_gates.Z)
+        # pauli_mask returns a value between 0 and 4 for [I, X, Y, Z].
```
Comment (balopat) [inline cirq-core/cirq/ops/clifford_gate.py:295]: where is `pauli_mask`?

File: cirq-core/cirq/ops/clifford_gate_test.py [hunk 5]
```diff
@@ -530,3 +530,25 @@ def test_to_phased_xz_gate(trans_x, trans_z):
...
+
+
+def test_clifford_tableau():
+    seen_tableau = []
+    for trans_x, trans_z in _all_rotation_pairs():
+        gate = cirq.SingleQubitCliffordGate.from_xz_map(trans_x, trans_z)
+        tableau_matrix = _extract_clifford_tableau_to_matrix(gate.clifford_tableau)
+        tableau_number = sum(2 ** i * t for i, t in enumerate(tableau_matrix.ravel()))
+        seen_tableau.append(tableau_number)
+        # Satify the symplectic property
```
Comment (balopat) [inline cirq-core/cirq/ops/clifford_gate_test.py:550]:
```suggestion
        # Satisfy the symplectic property
```

File: cirq-core/cirq/ops/clifford_gate_test.py [hunk 6]
```diff
@@ -448,7 +448,7 @@ def test_commutes_single_qubit_gate(gate, other):
     itertools.product(_all_clifford_gates(), _paulis, (1.0, 0.25, 0.5, -0.5)),
 )
 def test_commutes_pauli(gate, pauli, half_turns):
-    pauli_gate = pauli ** half_turns
+    pauli_gate = pauli if half_turns == 1 else pauli ** half_turns
```
Comment (balopat) [inline cirq-core/cirq/ops/clifford_gate_test.py:451]: Why is this needed? (pauli ** 1) should be equal to pauli

File: cirq-core/cirq/ops/clifford_gate_test.py [hunk 7]
```diff
@@ -530,3 +530,25 @@ def test_to_phased_xz_gate(trans_x, trans_z):
...
+    matrix[:, tableau.n * 2] = tableau.rs[: 2 * tableau.n]
+    return matrix
+
+
+def test_clifford_tableau():
+    seen_tableau = []
+    for trans_x, trans_z in _all_rotation_pairs():
+        gate = cirq.SingleQubitCliffordGate.from_xz_map(trans_x, trans_z)
+        tableau_matrix = _extract_clifford_tableau_to_matrix(gate.clifford_tableau)
+        tableau_number = sum(2 ** i * t for i, t in enumerate(tableau_matrix.ravel()))
```
Comment (balopat) [inline cirq-core/cirq/ops/clifford_gate_test.py:548]: This is nice but maybe we can simplify it further, e.g. the `_str_full_` might be cool too to use here as well? BTW - a digression - that `_str_full_` is something that I don't see being used anywhere else, I wonder why we wouldn't just use that as `_str_`?

**Review by balopat — CHANGES_REQUESTED** [review]
Looking good, with nits. Do we have any idea how much slower this representation will be compared to the current mapping?

File: cirq-core/cirq/ops/clifford_gate_test.py (hunk 6 above)
Comment (ybc1991) [reply cirq-core/cirq/ops/clifford_gate_test.py:451]: It is surprised to me as well. Actually, they return different type of classes:
```python
print(type(cirq.X), type(cirq.X**1))
# Prints:
# <class 'cirq.ops.pauli_gates._PauliX'> <class 'cirq.ops.common_gates.XPowGate'>
```

File: cirq-core/cirq/ops/clifford_gate.py (hunk 3 above)
Comment (ybc1991) [reply cirq-core/cirq/ops/clifford_gate.py:291]: In Single Qubit Clifford gate, it is not possible because it is automorphism (`I` always maps to `I`). So x_to and z_to should never map to `I`.
However, it is possible for multiple qubits case, say, we can map `XX` into  `XI`.

File: cirq-core/cirq/ops/clifford_gate.py (hunk 1 above)
Comment (ybc1991) [reply cirq-core/cirq/ops/clifford_gate.py:46]: Thanks for catching this. It is embarrassing to has this typo. (Relying too much on the auto-fill, lol). Fixed all other 12 places :)

File: cirq-core/cirq/ops/clifford_gate.py (hunk 2 above)
Comment (ybc1991) [reply cirq-core/cirq/ops/clifford_gate.py]: Done

File: cirq-core/cirq/ops/clifford_gate.py (hunk 4 above)
Comment (ybc1991) [reply cirq-core/cirq/ops/clifford_gate.py:295]: the line below: `to_gate = single_pauli_gates[to.pauli_mask[0] - 1]` in the middle of getting index.
Just realized it can also be simplified into `to_gate = Pauli._XYZ[to.pauli_mask[0] - 1]`

File: cirq-core/cirq/ops/clifford_gate_test.py (hunk 7 above)
Comment (ybc1991) [reply cirq-core/cirq/ops/clifford_gate_test.py:548]: I do use `_str_full_`several times during the debugging. I just feel using string in test code is harder to manipulate compared with array. This code is a little bit out-of-date, actually. I have created `matrix` method in the tableau class. PTAL. But if you have better ideas, I'd love to update.

**Comment by ybc1991** [comment]
> Looking good, with nits. Do we have any idea how much slower this representation will be compared to the current mapping?

Thanks for reviewing. This is just single qubit clifford gate. One more or two more dictionary lookups probably are negligible. 
Plus, those extract lookup should only happen once to construct the gate. The repeated steps involve the methods like either `_unitary_` or `_decompose_` is untouched. This class has no `_act_on_` yet. But having tableau won't slow down it.

**Comment by 95-martin-orion** [comment]
Pinging @balopat for re-review.

**Comment by ybc1991** [comment]
@balopat Could you review this PR again? This one is kind of blocking of following General Clifford Gate PRs. Thanks.

File: cirq-core/cirq/ops/clifford_gate.py [hunk 8]
```diff
@@ -43,6 +43,36 @@ def _to_pauli_transform(matrix: np.ndarray) -> Optional[PauliTransform]:
...
+def _to_clifford_tableau(
+    rotation_map: Optional[Dict[Pauli, PauliTransform]] = None,
+    *,
+    x_to: Optional[PauliTransform] = None,
+    z_to: Optional[PauliTransform] = None,
+) -> qis.CliffordTableau:
+    """Transfer the rotation map to clifford tableau representation"""
+    if x_to is None and z_to is None and rotation_map is None:
+        # coverage: ignore
+        raise ValueError(
```
Comment (balopat) [inline cirq-core/cirq/ops/clifford_gate.py:55]: Let's cover this with tests instead of ignoring it, error scenarios are as important as happy path scenarios.

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 9]
```diff
@@ -15,7 +15,8 @@
 from typing import Any, Dict, List
 import numpy as np
 
-import cirq
+# import cirq is needed for mypy
```
Comment (balopat) [inline cirq-core/cirq/qis/clifford_tableau.py:18]: we typically do this differently: 

```
if TYPE_CHECKING:
   import cirq
```

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 10]
```diff
@@ -306,6 +307,8 @@ def _row_to_dense_pauli(self, i: int) -> 'cirq.DensePauliString':
             represents the effective single Pauli operator on that qubit. The
             overall phase is captured in the coefficient.
         """
+        from cirq.ops.dense_pauli_string import DensePauliString
```
Comment (balopat) [inline cirq-core/cirq/qis/clifford_tableau.py:310]: we try to avoid inline imports, instead use `from cirq import ops` on the top and here you can use `ops.DensePauliString` to avoid circular import issues.

File: cirq-core/cirq/ops/clifford_gate_test.py (hunk 6 above)
Comment (balopat) [reply cirq-core/cirq/ops/clifford_gate_test.py:451]: Let's add a TODO with https://github.com/quantumlib/Cirq/issues/4328 on it for this

**Review by balopat — APPROVED** [review]
LGTM with nits

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 9 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:18]: Done

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 10 above)
Comment (ybc1991) [reply cirq-core/cirq/qis/clifford_tableau.py:313]: Hmmm, if we move it to the top, I still encounter the error:

```
ImportError while loading conftest '/Users/ybc/Documents/github/MyCirq/cirq-core/cirq/conftest.py'.
cirq-core/cirq/__init__.py:30: in <module>
    from cirq import (
cirq-core/cirq/qis/__init__.py:24: in <module>
    from cirq.qis.clifford_tableau import CliffordTableau
cirq-core/cirq/qis/clifford_tableau.py:18: in <module>
    from cirq import ops
cirq-core/cirq/ops/__init__.py:21: in <module>
    from cirq.ops.clifford_gate import (
cirq-core/cirq/ops/clifford_gate.py:51: in <module>
    ) -> qis.CliffordTableau:
E   AttributeError: module 'cirq.qis' has no attribute 'CliffordTableau'
```
Looks like circular import again.

File: cirq-core/cirq/ops/clifford_gate.py (hunk 8 above)
Comment (ybc1991) [reply cirq-core/cirq/ops/clifford_gate.py:55]: Done.

File: cirq-core/cirq/ops/clifford_gate_test.py (hunk 6 above)
Comment (ybc1991) [reply cirq-core/cirq/ops/clifford_gate_test.py:452]: Done.

**Comment by ybc1991** [comment]
Hi @MichaelBroughton, as we have talked about in the last week, could you take a look at this SingleCliffordGate PR? Thanks!

**Review by MichaelBroughton — APPROVED** [review]

**Comment by MichaelBroughton** [comment]
@ybc1991 it looks like in the time this PR has spent open, it has slipped on a few of the integration tests. Would you mind fixing those up and then we can merge this in ASAP ?

**Comment by ybc1991** [comment]
> @ybc1991 it looks like in the time this PR has spent open, it has slipped on a few of the integration tests. Would you mind fixing those up and then we can merge this in ASAP ?

Yes, I have updated the file. I think the error was because of the serialization, which was based on the old rotation and inverse map, which won't exist as the class members after this PR. PTAL again.

**Comment by MichaelBroughton** [comment]
This looks good. Can you try commenting the @googlebot I fixed it to see if that can flip the switch on this CLA check ?

**Comment by ybc1991** [comment]
@googlebot I fixed it.

**Comment by ybc1991** [comment]
@googlebot I fixed it.
